### PR TITLE
Record stream deployment

### LINKED ISSUE #1681: Record stream deployments
As a developer, I'd like to establish the core schema changes to persist stream deployments.

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Add StreamDeployment repository that keeps track stream deployments
   - Define necessary operations in StreamDeploymentRepository interface
   - Implement both RDBMS and in-memory (for test only)
   - Define DDL for supported DB schemas
 - Add StreamDeployment domain object
 - Verify StreamDeployment entry when upgrading stream
 - Verify StreamDeployment when calculating state
 - Remove usage of USE_SKIPPER as the system property
 - Add tests to verify both the upgrade and state in DefaultStreamService

Resolves #1681

### CODE REVIEW

**Comment by ilayaperumalg** [comment]
Squashed and rebased.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultStreamService.java [hunk 1]
```diff
@@ -416,17 +426,45 @@ private boolean isPartitionedConsumer(Map<String, String> appDeploymentPropertie
...
+			}
+			else {
+				if (streamDeployment.getDeployerName().equals(StreamDeployers.skipper.name())) {
+					skipperStreams.add(streamDefinition);
+				}
+				else if (streamDeployment.getDeployerName().equals(StreamDeployers.appdeployer.name())){
+					appDeployerStreams.add(streamDefinition);
+				}
+				else {
+					logger.error("Unknown deployer " + streamDeployment.getDeployerName() +
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultStreamService.java]: Just wondering if we should throw an exception here.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java [hunk 2]
```diff
@@ -184,18 +192,23 @@ public void deployStream(StreamDeploymentRequest streamDeploymentRequest) {
...
 		packageIdentifier.setPackageVersion("1.0.0");
-		packageIdentifier.setRepositoryName("local");
+		String repoName = "local";
+		packageIdentifier.setRepositoryName(repoName);
 		installRequest.setPackageIdentifier(packageIdentifier);
 		InstallProperties installProperties = new InstallProperties();
 		installProperties.setPlatformName("default");
 		//todo: Better naming for the release name
-		installProperties.setReleaseName("my" + streamDeploymentRequest.getStreamName());
+		String releaseName = "my" + streamName;
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java]: Should `my` be something like `SCDF`?

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentControllerTests.java [hunk 3]
```diff
@@ -81,10 +82,14 @@
 	@Mock
 	private DefaultStreamService defaultStreamService;
 
+	@Mock
+	private StreamDeploymentRepository stramDeploymentRepository;
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/controller/StreamDeploymentControllerTests.java]: `streamDeploymentRepository` is misspelled.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/repository/InMemoryStreamDeploymentRepository.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2015 the original author or authors.
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/repository/InMemoryStreamDeploymentRepository.java]: 2017 ;-)

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/services/impl/DefaultStreamServiceTests.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/services/impl/DefaultStreamServiceTests.java]: This is a new class so just 2017.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/services/impl/DefaultStreamServiceTests.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+		StreamDeployment streamDeployment3 = new StreamDeployment(streamDefinition3.getName(),
+				StreamDeployers.skipper.name(), "pkg1", "release2", "local");
+		when(streamDeploymentRepository.findOne(streamDeployment1.getStreamName())).thenReturn(streamDeployment1);
+		when(streamDeploymentRepository.findOne(streamDeployment2.getStreamName())).thenReturn(streamDeployment2);
+		when(streamDeploymentRepository.findOne(streamDeployment3.getStreamName())).thenReturn(streamDeployment3);
+	}
+
+	@Test
+	public void verifyUpgradeStream() {
+		StreamDeploymentRepository streamDeploymentRepository = mock(StreamDeploymentRepository.class);
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/services/impl/DefaultStreamServiceTests.java]: Why aren't we using the mocks we setup in the `@Before`?

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/services/impl/DefaultStreamServiceTests.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+				null, null);
+		verify(skipperStreamDeployer, times(1)).upgradeStream(streamDeployment2.getStreamName(),
+				streamDeployment2.getReleaseName(), null, null);
+		verifyNoMoreInteractions(skipperStreamDeployer);
+		verify(appDeployerStreamDeployer, never()).deployStream(any());
+	}
+
+	@Test
+	public void verifyAppDeployerUpgrade() {
+		StreamDeploymentRepository streamDeploymentRepository = mock(StreamDeploymentRepository.class);
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/services/impl/DefaultStreamServiceTests.java]: Why aren't we using the mocks we setup in the `@Before`?

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultStreamService.java (hunk 1 above)
Comment (ilayaperumalg) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultStreamService.java]: Currently, we have the pre-defined deployer names and it is most likely to have this path of the code to execute. Given this is for state() calculation and throwing an exception for one incorrect entry may not be necessary. Hence, just error logging.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java (hunk 2 above)
Comment (ilayaperumalg) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/stream/SkipperStreamDeployer.java]: We have a story to cover this: https://github.com/spring-cloud/spring-cloud-dataflow/issues/1688

**Comment by ilayaperumalg** [comment]
@cppwfs thank you for the review. Pushed the changes that address review comments.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultStreamService.java (hunk 1 above)
Comment (cppwfs) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultStreamService.java]: Understood.   I guess the log is enough, just was concerned about leaving an entry out of the result.

**Comment by cppwfs** [comment]
I'll kick the tires a bit more.   But would like @jvalkeal or @markpollack to take give it a once over before merging.

**Comment by jvalkeal** [comment]
@cppwfs I think from my side this looks ok.

**Comment by cppwfs** [comment]
👍

**Comment by cppwfs** [comment]
LGTM

**Comment by cppwfs** [comment]
Squashed Merged.

**Pull request closed without merging**